src/Export/ProductService.php

```diff
@@ -261,4 +244,47 @@ protected function addProductIdFilters(Criteria $criteria, string $productId): v
             )
         );
     }
+
+    protected function getVisibleProducts(?int $limit, ?int $offset, ?string $productId): EntitySearchResult
+    {
+        $criteria = $this->getCriteriaWithProductVisibility($limit, $offset);
+
+        if ($productId) {
+            $this->addProductIdFilters($criteria, $productId);
+        }
+
+        return $this->container->get('product.repository')->search(
+            $criteria,
+            $this->salesChannelContext->getContext()
+        );
+    }
+
+    protected function buildProductsWithVariantInformation(EntitySearchResult $result): ProductCollection
+    {
+        $products = new ProductCollection();
+
+        /** @var ProductEntity $product */
+        foreach ($result->getEntities() as $product) {
+            if ($product->getMainVariantId() !== null) {
+                $product = $this->getRealMainProductWithVariants($product->getMainVariantId());
+            } else {
+                $children = $this->getChildrenOrSiblings($product);
+                $product->setChildren($children);
+            }
+
+            $products->add($product);
+        }
+
+        return $products;
+    }
+
+    protected function getRealMainProductWithVariants(string $realMainProductId): ProductEntity
+    {
+        $newProduct = $this->getVisibleProducts(1, 0, $realMainProductId)->first();
+
+        $children = $this->getChildrenOrSiblings($newProduct);
+        $newProduct->setChildren($children);
```

Comment:
Those two lines are also used in buildProductsWithVariantInformation and could be removed here to avoid duplicate code.